src/site/stages/build/plugins/modify-dom/index.js

```diff
@@ -0,0 +1,47 @@
+/* eslint-disable no-param-reassign */
+const path = require('path');
+const cheerio = require('cheerio');
+const addNonceToScripts = require('./add-nonce-to-scripts');
+const processEntryNames = require('./process-entry-names');
+const updateExternalLinks = require('./update-external-links');
+const addSubheadingsIds = require('./add-id-to-subheadings');
+const checkBrokenLinks = require('./check-broken-links');
+const injectAxeCore = require('./inject-axe-core');
+
+const domModifiers = [
+  addNonceToScripts,
+  processEntryNames,
+  updateExternalLinks,
+  addSubheadingsIds,
+  checkBrokenLinks,
+  injectAxeCore,
+];
+
+const parseHtml = BUILD_OPTIONS => files => {
+  for (const modifier of domModifiers) {
+    if (modifier.initialize) {
+      modifier.initialize(BUILD_OPTIONS, files);
+    }
+  }
+
+  for (const [fileName, file] of Object.entries(files)) {
+    if (path.extname(fileName) === '.html') {
+      file.dom = cheerio.load(file.contents);
+      for (const modifier of domModifiers) {
+        modifier.modifyFile(fileName, file, files, BUILD_OPTIONS);
+      }
+      if (file.modified) {
+        file.contents = Buffer.from(file.dom.html());
+      }
+      delete file.dom;
+    }
+  }
+
+  for (const modifier of domModifiers) {
+    if (modifier.conclude) {
+      modifier.conclude(BUILD_OPTIONS, files);
+    }
+  }
+};
+
+module.exports = parseHtml;
```

Comment:
This whole plugin is the key change. Everything listed in the `domModifiers` list use to be a separate Metalsmith plugin. This PR converts them into sub-steps within this plugin where each sub-step instead operates on a single file at a time, so that only one `file.dom` exists in memory at once.